lib/gen_stage.ex

```diff
@@ -1027,45 +1087,81 @@ defmodule GenStage do
-        stage = put_in stage.producers[ref], {producer_pid, cancel, max, min, max}
-        {{:ok, ref}, stage}
-      else
-        ref = make_ref()
-        stage = put_in stage.producers[ref], {to, cancel, max, min, max}
-        send self(), {:DOWN, ref, :process, to, :noproc}
-        {{:ok, ref}, stage}
-      end
+      producer_pid = GenServer.whereis(to)
+      cond do
+        producer_pid != nil ->
+          ref = Process.monitor(producer_pid)
+          send producer_pid, {:"$gen_producer", {self(), ref}, {:subscribe, opts}}
+          handle_subscribe(full_opts, ref, producer_pid, cancel, {max, min, max}, stage)
+        cancel == :temporary ->
+          {:reply, {:ok, make_ref()}, stage, :infinity}
+        cancel == :permanent ->
+          {:stop, :noproc, {:ok, make_ref()}, stage}
+       end
     else
       {:error, message} ->
         :error_logger.error_msg('GenStage subscribe received invalid option: ~ts~n', [message])
-        {{:error, {:bad_opts, message}}, stage}
+        {:reply, {:error, {:bad_opts, message}}, stage, :infinity}
+    end
+  end
+
+  defp handle_subscribe(opts, ref, producer_pid, cancel, demand, stage) do
+    to = {producer_pid, ref}
+    case handle_subscribe(opts, to, stage) do
+      {:automatic, stage, timeout} ->
+        {ask, _, _} = demand
+        ask(to, ask)
+        stage = put_in stage.producers[ref], {producer_pid, cancel, demand}
+        {:reply, {:ok, ref}, stage, timeout}
+      {:manual, stage, timeout} ->
+        stage = put_in stage.producers[ref], {producer_pid, cancel, :manual}
+        {:reply, {:ok, ref}, stage, timeout}
+      {:stop, _, _} = stop ->
+        stop
+    end
+  end
+
+  defp handle_subscribe(opts, to, %{mod: mod, state: state} = stage) do
+    case apply(mod, :handle_subscribe, [opts, to, state]) do
+      {mode, state} when mode in [:automatic, :manual] ->
+        {mode, %{stage | state: state}, :infinity}
+      {mode, state, timeout} when mode in [:automatic, :manual] ->
+        {mode, %{stage | state: state}, timeout}
+      {:stop, reason, state} ->
+        {:stop, reason, %{stage | state: state}}
     end
   end
 
   defp cancel_producer(ref, reason, %{producers: producers} = stage) do
     case Map.pop(producers, ref) do
       {nil, _producers} ->
         {:noreply, stage}
-      {{_, :temporary, _, _, _}, producers} ->
+      {{_, :temporary, {_, _, _}}, producers} ->
         Process.demonitor(ref, [:flush])
         {:noreply, %{stage | producers: producers}}
-      {{_, :permanent, _, _, _}, producers} ->
+      {{producer_pid, :temporary, :manual}, producers} ->
+        handle_cancel(reason, {producer_pid, ref}, %{stage | producers: producers})
+      {{_, :permanent, _}, producers} ->
         Process.demonitor(ref, [:flush])
         {:stop, reason, %{stage | producers: producers}}
     end
   end
 
+  defp handle_cancel(reason, to, %{mod: mod, state: state} = stage) do
```

Comment:
Why don't we use the regular handle_callback? If someone wants to dispatch events on cancel... that's fine. At least the API is consistent?